src/batch.js

```diff
@@ -1,28 +1,221 @@
+
 const { HttpError } = require('./httpError')
 
 const batchDefaultConfig = {
-  maxBatchSize: 20
+  maxBatchSize: 20,
+  handler: undefined
 }
 
 module.exports = {
   batchHandler
 }
 
-async function batchHandler({ routes, config, extractPathParameters }, event, context) {
+async function batchHandler({ route, config }, event, context) {
   config = { ...batchDefaultConfig, ...config }
+  const body = event.body
 
   // Validate
-  // validate json body has been parsed
-  // validate body.requests has array of requests
-  // validate each body.requests[]
+  validateBatchRequest(body, config.maxBatchSize)
 
   // build graph of requests
+  const orderedRequests = tSort(body.requests)
+  const { asyncRequests, syncRequests } = splitAsyncRequests(orderedRequests)
 
   // execute requests
+  const results = await Promise.all([
+    (async () => {
+      const syncResults = []
+      for (let req of syncRequests) {
+        syncResults.push(await executeRequest(config, route, event, req))
+      }
+      return syncResults
+    })(),
+    ...asyncRequests.map(executeRequest.bind(null, config, route, event))
+  ])
 
   // return response
+  return {
+    statusCode: 200,
+    body:{ responses: results.map((result, index) => {
+      if (index === 0){
+        return result.map((result, index) => {
+          return {
+            id: syncRequests[index].id,
+            status: result.statusCode,
+            body: result.body ? JSON.parse(result.body) : undefined,
+            headers: result.headers
+          }
+        })
+      }
+      return {
+        id: asyncRequests[index - 1].id,
+        status: result.statusCode,
+        body: result.body ? JSON.parse(result.body) : undefined,
+        headers: result.headers
+      }
+    })},
+    headers:{}
+  }
 }
 
-function validateBatchRequest(body) {}
+async function executeRequest(config, route, event, request) {
+  //Parse url into pathParameters and multiValueQueryStringParameters formats
+  const urlAndQuerystring = request.url.split('?')
+  let queryStringParameters = {}
+  if (urlAndQuerystring.length > 1) {
+    const queryStringParams = urlAndQuerystring[1]
+    const keys = queryStringParams.split('&')
+    const queryStrings = keys.map((k) => k.split('='))
+
+    for (let queryString of queryStrings) {
+      if (queryStringParameters[queryString[0]]) {
+        queryStringParameters[queryString[0]].push(queryString[1])
+      } else {
+        queryStringParameters[queryString[0]] = [queryString[1]]
+      }
+    }
+  }
+
+  //Compose event to appear as native AWS event
+  let childEvent = { 
+    ...event,
+    httpMethod: request.method,
+    headers: {...normalizeRequestHeaders(request.headers), authorization: event.headers.authorization},
+    body: request.body,
+    path: request.url,
+    pathParameters: { 
+      proxy: urlAndQuerystring[0].split('/').splice(2).join('/') 
+    },
+    multiValueQueryStringParameters: queryStringParameters,
+    _json: true
```

Comment:
What is this property for?

Reply:
multiValueQueryStringParameters is something introduced a couple years back to fix the issue in API Gateway where you'd only get one of the values when you have querystrings with the same name.

Reply:
Yep, I'm referring to the `_json` property.

Reply:
Oop, yeah of course you were.

I think it's probably from the lambda tester, SAM, or lambda-local as most of these values were originally selected months back from examining a local event.

I've removed it there's some reason for it being set, it should already be on the event.